src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/WellKnownTypes.cs

```diff
@@ -0,0 +1,187 @@
+        "Microsoft.AspNetCore.Mvc.ControllerAttribute",
+        "Microsoft.AspNetCore.Mvc.NonControllerAttribute",
+        "Microsoft.AspNetCore.Mvc.NonActionAttribute",
+        "Microsoft.AspNetCore.Http.AsParametersAttribute",
+        "System.Threading.CancellationToken",
+        "Microsoft.AspNetCore.Http.HttpContext",
+        "Microsoft.AspNetCore.Http.HttpRequest",
+        "Microsoft.AspNetCore.Http.HttpResponse",
+        "System.Security.Claims.ClaimsPrincipal",
+        "Microsoft.AspNetCore.Http.IFormFileCollection",
+        "Microsoft.AspNetCore.Http.IFormFile",
+        "System.IO.Stream",
+        "System.IO.Pipelines.PipeReader",
+        "System.IFormatProvider",
+        "System.Uri",
+        "Microsoft.AspNetCore.Builder.ConfigureHostBuilder",
+        "Microsoft.AspNetCore.Builder.ConfigureWebHostBuilder",
+        "Microsoft.Extensions.Hosting.GenericHostWebHostBuilderExtensions",
+        "Microsoft.AspNetCore.Hosting.WebHostBuilderExtensions",
+        "Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions",
+        "Microsoft.Extensions.Hosting.HostingHostBuilderExtensions",
+        "Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions",
+        "Microsoft.AspNetCore.Builder.WebApplication",
+        "Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions",
+        "System.Delegate",
+        "Microsoft.AspNetCore.Mvc.ModelBinding.IBinderTypeProviderMetadata",
+        "Microsoft.AspNetCore.Mvc.BindAttribute",
+        "Microsoft.AspNetCore.Http.IResult",
+        "Microsoft.AspNetCore.Mvc.IActionResult",
+        "Microsoft.AspNetCore.Mvc.Infrastructure.IConvertToActionResult",
+        "Microsoft.AspNetCore.Http.RequestDelegate",
+        "System.Threading.Tasks.Task`1",
+    };
+    
+    public static WellKnownTypes GetOrCreate(Compilation compilation) =>
+        _compilationToLazyWellKnownTypes.GetValue(compilation, static c => new WellKnownTypes(c));
+    
+    private readonly INamedTypeSymbol?[] _lazyWellKnownTypes;
+    private readonly Compilation _compilation;
+
+    static WellKnownTypes()
+    {
+        AssertEnumAndTableInSync();
+    }
+    
+    [Conditional("DEBUG")]
+    private static void AssertEnumAndTableInSync()
+    {
+        for (var i = 0; i < _wellKnownTypeNames.Length; i++)
+        {
+            var name = _wellKnownTypeNames[i];
+            var typeId = (WellKnownType)i;
+
+            var typeIdName = typeId.ToString().Replace("__", "+").Replace('_', '.');
+
+            var separator = name.IndexOf('`');
+            if (separator >= 0)
+            {
+                // Ignore type parameter qualifier for generic types.
+                name = name.Substring(0, separator);
+                typeIdName = typeIdName.Substring(0, separator);
+            }
+
+            Debug.Assert(name == typeIdName, $"Enum name ({typeIdName}) and type name ({name}) must match at {i}");
+        }
+    }
```

Comment:
💡 There is no need to run this on every client machine with every use. It can be moved to a unit test.

Reply:
One benefit of validating here is that it will cache errors when someone changes this type and then just runs their particular test. I guess that is more common than rerunning all tests after a change.